Jack cut feature

chenkasirer opened the pull request:
* Example implementation of a merged `JackRafterCut` and the `CutFeature`. Including client code in `L-Miter`.

I'm making an effort to make this change as gradual as possible. Meaning the two systems live side-by-side. For that purpose I created the new `JackRafterCut` in a new package `_fabrication` as otherwise there are some circular import issues.

![image](https://github.com/user-attachments/assets/baca47f2-1b0f-4a7a-8909-5bbd880dec58)

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ ] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [ ] I added new functions/classes and made them available on a second-level import, e.g. `compas_timber.datastructures.Beam`.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added necessary documentation (if appropriate)

papachap commented on src/compas_timber/elements/beam.py:
```diff
@@ -467,7 +485,28 @@ def remove_blank_extension(self, joint_key=None):
+            The index of the reference side to be returned. 0 to 5.
+
+        """
+        # TODO: maybe this should be the default representation of the ref sides?
```
I believe so as well. Using a parametric planar surface for the ref_side makes it more comprehensible i think.

chenkasirer replied:
alright then I'd keep a mental note of maybe removing the frame-like ref sides. and we get rid of them in the future

obucklin replied:
this comment was clarified in another PR, you can ignore my comment

papachap commented on src/compas_timber/elements/beam.py:
```diff
@@ -406,6 +420,7 @@ def _create_shape(frame, xsize, ysize, zsize):
 
     @reset_computed
```
This resets the cached attributes of the class?

chenkasirer replied:
yes. some `Element` has some internal instance attributes which are cached after the first time they are created. For example `Element._geometry` is one. These need to be cleared when changes are made which might result in new geometry. This decorator takes care of that.

papachap commented on src/compas_timber/_fabrication/jack_cut.py:
```diff
@@ -0,0 +1,276 @@
+    def orientation(self):
+        return self._orientation
+
+    @orientation.setter
```
what is the benefit of using setters and private properties for this class?

chenkasirer replied:
this an object oriented thing that's strongly related to the encapsulation principle. In statically typed languages where private access is enforced you'd typically always need setters and getters. In python, since nothing is enforced by the language, it can get blurry.

There's different reasons why you'd want to use properties in python. In this particular case, I used properties and their setters for input validation. Since BTLx has a clear specification for the allowed value ranges for each parameter it seemed like a good place to validate those.

papachap commented on src/compas_timber/_fabrication/jack_cut.py:
```diff
@@ -0,0 +1,276 @@
+    END = 1
+
+
+class JackRafterCut(BTLxProcess):
```
Where does the serialization of the btlx process parameters happen? shouldn't there be a class method for that? Also does this mean that header attributes are being removed?

chenkasirer replied:
You're right, that's missing. I introduced this in one of the last commits, might have not pushed yet.

I'm still very much debating though. I'd like to extract the XML schema specific stuff out of these classes, but they are nevertheless related. In the meantime, I created a `JackRafterCutParams` class whose sole responsibility is to populate the parameter values from `JackRafterCut` in they appropriate key-value pair according to the specification.

chenkasirer replied:
it's in the latest commit https://github.com/gramaziokohler/compas_timber/pull/267/commits/aec27952f88eb1eb127eb97cfde185a936d68ada

papachap commented on src/compas_timber/connections/l_miter.py:
With these changes does the l_mitter factory file gets removed?

chenkasirer replied:
yup. that's the intention. `LMiterJoint` becomes the de-facto factory as it directly assigns the processes/features to the respective beams.

papachap commented:
In case there are two btlx processes that are correlated (i.e StepJoint, StepJointNotch) and share most most of the parameters should the process be implemented as:

- two separate classes in the same module (_fabrication/step_joint) and generate the parameters twice if they both need to be applied in the case of a joint with two beams (one on main_beam and one cross_beam).

or


- as one class containing all the combined parameters with a boolean for applying the step or the step_notch

_(I attached the parameters for each one of the processes for the example of the StepJoint/StepJointNotch)_
![Screenshot 2024-07-23 102032](https://github.com/user-attachments/assets/0f5f2ab5-ed2a-44d8-bff9-42e05da4263a)
![Screenshot 2024-07-23 102229](https://github.com/user-attachments/assets/de17720f-30ce-476d-95ff-fd5f1ebc8047)

papachap commented on src/compas_timber/_fabrication/jack_cut.py:
```diff
@@ -0,0 +1,274 @@
+        start_x = distance_point_point(ref_edge.point, point_start_x)
+        angle = cls._calculate_angle(ref_side, plane, orientation)
+        inclination = cls._calculate_inclination(ref_side, plane, orientation)
+        return cls(orientation, start_x, start_y, start_depth, angle, inclination, ref_side_index=ref_side_index)
```
should the ref_side_index be in the this class constructor, since it is not a parameter of the class?

chenkasirer replied:
`ref_side_index` is an attribute of the parent class, as all processes need this parameter. notice `JackRafterCut` knows nothing about it, it gets it as a `kwarg` and passes it through to the parent in its call to `super().__init__()`.

chenkasirer commented on src/compas_timber/elements/beam.py:
```diff
@@ -252,7 +265,7 @@ def __str__(self):
     # ==========================================================================
 
     def compute_geometry(self, include_features=True):
```
exactly. `Beam` now creates its own geometry as a result as its `blank` + features.

chenkasirer commented:
> In case there are two btlx processes that are correlated (i.e StepJoint, StepJointNotch) and share most of the parameters, should the process be implemented as:
> 
> * two separate classes in the same module (_fabrication/step_joint) and generate the parameters twice if they both need to be applied in the case of a joint with two beams (one on main_beam and one cross_beam).
> 
> or
> 
> * as one class containing all the combined parameters with a boolean for applying the step or the step_notch
> 
> _(I attached the parameters for each one of the processes for the example of the StepJoint/StepJointNotch the common parameters should have the same values if a joint between two beams is created)_ ![Screenshot 2024-07-23 102032](https://private-user-images.githubusercontent.com/145543755/351245164-0f5f2ab5-ed2a-44d8-bff9-42e05da4263a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE4MTQxMzcsIm5iZiI6MTcyMTgxMzgzNywicGF0aCI6Ii8xNDU1NDM3NTUvMzUxMjQ1MTY0LTBmNWYyYWI1LWVkMmEtNDRkOC1iZmY5LTQyZTA1ZGE0MjYzYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcyNFQwOTM3MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hZTNmNjBmYjI3MDA5Zjk0Y2M0ZGI5MzdjMzA3NDljOTE4Zjc5NjkwZTE3NzY0ODUzNWI5YmM4YzAzNGY3ZjhjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.mwxhdXC8WWHgiM43hDWHOHHPpv8t82WhqgRAHzlrH6k) ![Screenshot 2024-07-23 102229](https://private-user-images.githubusercontent.com/145543755/351245668-de17720f-30ce-476d-95ff-fd5f1ebc8047.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjE4MTQxMzcsIm5iZiI6MTcyMTgxMzgzNywicGF0aCI6Ii8xNDU1NDM3NTUvMzUxMjQ1NjY4LWRlMTc3MjBmLTMwY2UtNDc2ZC05NWZmLWZkNWYxZWJjODA0Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNzI0JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDcyNFQwOTM3MTdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01ODJmNTMyNzJjMTFiYzRmNjYzYjVlMzVkNjRmZjAyYWI2MjdhZjI3NWZhNDBhMzVlZGJkYzcyNTM5MGJkZTllJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.x1mZBWpCrxsvliFEbPfHmxlTmQ8492PmL5oD5IMz0SM)

here I'd implement both separately. could be in the same module, but definitely as two separate classes. If the parameters of one of them is a subset of the other's then you could use inheritance to avoid having duplication code. 

Additionally, if it makes sense, we could make a factory method which creates matching pairs of `StepJoint` and `StepJointNotch` from the necessary collection of parameters.

chenkasirer commented on src/compas_timber/_fabrication/jack_cut.py:
```diff
@@ -0,0 +1,276 @@
+        if point_start_x is None:
+            raise ValueError("Plane does not intersect with beam.")
+
+        start_x = distance_point_point(ref_edge.point, point_start_x)
```
why? :)

papachap replied:
ref_edge is a compas Line, no? in this case shouldn't it be ref_edge.start? I thought it was a typo or smth ;p

chenkasirer commented:
@papachap I think I addressed all of your comments. 
I think it is currently not necessary to partially replace the processes just for the L-Miter. I'd leave it like this for now and make the swap from joint factories to process/features once we have all processes properly implemented.

If it's all the same to you, I'd go ahead an merge this one. Just hit approve if it's all good, otherwise feel free to leave some more comments.

papachap reviewed (approved):
@chenkasirer Perfect, thanks! This was helpful! I'll try to do a pull request before next week for the `StepJoint` and `StepJointNotch` based on your implementation of the `JackRafterCut` and your feedback. We can discuss on that once I have it :)

obucklin reviewed (commented):
generally looks good. I think we just need to test the generation of the two angles with various plane angles.
